### DISCUSSION TITLE
How to unit test a method which is decorated with an aspect that inject a dependency?

### DISCUSSION THREAD

**Discussion by MEmanuelsson** [metalama/Metalama#234] [Q&A] [answered]
I've tried to find an answer to this problem in documentation and previous issues here but without any luck. 

I have an aspect which injects a particular dependency (IDistributedCache) using [IntroduceDependency]-attribute. 

This aspect is added to a function in a class, in this case called ContentService like this (simplifed example):

```
public class ContentService : IContentService
{
    private readonly ILogger<ContentService> _logger;
    
    public ContentService(ILogger<ContentService> logger)
    {
        _logger = logger;
    }

    [Cache]
    public async Task<SiteContent> GetSiteContent(string host)
    {
    }
}
```

Then in my unit test I instantiate an object using a mocked version of ILogger like this (in short):

`_service = new ContentService(_l.Object);`

However this call throws an exception since ContentService now also expects to have IDistributedCache in the constructor. 

How can I fix this issue? Is it possible to disable the aspect in unit tests, or inject IDistributedCache in some way?
/Martin

**Comment by gfraiteur** [comment, marked as answer]
It is not yet supported. You need to build your instance using the DI container, i.e. through reflection.

**Reply by MEmanuelsson** [reply]
Thanks a lot, that works like a charm 👍
